File: Algorithms/ShortestPath/BellmanFords.cpp
#include<bits/stdc++.h>
using namespace std;

struct Edge {
    int source;
    int dest;
    int weight;
};

void printPath(vector<int> const &parent, int v) {
    if (v < 0)  return;

    printPath(parent, parent[v]);
    cout << v << " ";
}

void BellmanFord(vector<Edge> const &edges, int source, int N) {
    int E = edges.size();

    vector<int> distance (N, INT_MAX);
    distance[source] = 0;

    vector<int> parent (N, -1);

    int u,v, w, K=N;
    while(--K) {
        for(int j=0; j<E; j++) {
            u = edges[j].source;
            v = edges[j].dest;
            w = edges[j].weight;

            if (distance[u] != INT_MAX && distance[u]+w < distance[v]) {
                distance[v] = distance[u] + w;
                parent[v]   = u;
            }
        }
    }

    /// Check for negative weight cycles
    for(int i=0; i<E; i++) {
        u = edges[i].source;
        v = edges[i].dest;
        w = edges[i].weight;

        if (distance[u] != INT_MAX && distance[u]+w < distance[v]) {
            cout << "Negative weight cycle Found!!\n";
            return;
        }
    }

    for(int i=0; i<N; i++) {
        cout << "Distance of vertex " << i << " from the source is " << setw(2) << distance[i] << ". It's path is[ " ;
        printPath(parent, i);
        cout << "]" << endl;
    }
}

int main() {
    vector<Edge> edges = {
        /// (u,v, w) -> edge from u to v having weight w
        {0, 1,-1},  {0, 2, 4}, {1, 2, 3},  {1, 3, 2},
        {1, 4, 2},  {3, 2, 5}, {3, 1, 1},  {4, 3,-3}
    };

    int source = 0;
    int N = 5;
    BellmanFord(edges, source, N);

    return 0;
}

/**
OUTPUT:
        Distance of vertex 0 from the source is  0. It's path is[ 0 ]
        Distance of vertex 1 from the source is -1. It's path is[ 0 1 ]
        Distance of vertex 2 from the source is  2. It's path is[ 0 1 2 ]
        Distance of vertex 3 from the source is -2. It's path is[ 0 1 4 3 ]
        Distance of vertex 4 from the source is  1. It's path is[ 0 1 4 ]

**/

File: Algorithms/ShortestPath/BellmanFords.java
package ShortestPath;

import java.util.*;

class Edge {
	int source;
    int dest;
    int weight;
    
    public Edge(int source, int dest, int weight) {
    	this.source = source;
    	this.dest   = dest;
    	this.weight = weight;
    }
}

public class BellmanFords {
	
	public static void printPath(int parent[], int v) {
		if (v < 0)	return;
		
		printPath(parent, parent[v]);
		System.out.print(v + " ");
	}
	
	public static void BellmanFord(List<Edge> edges, int source, int N) {
		int E = edges.size();
		
		int distance[] = new int[N];
		int parent[]   = new int[N];
		
		Arrays.fill(distance, Integer.MAX_VALUE);
		distance[source] = 0;
		
		Arrays.fill(parent, -1);
		int K = N;
		while (--K > 0) {
			for(int j=0; j<E; j++) {
				int u = edges.get(j).source;
				int v = edges.get(j).dest;
				int w = edges.get(j).weight;
				
				if (distance[u] != Integer.MAX_VALUE && (distance[u]+w) < distance[v]) {
					distance[v] = distance[u] + w;
					parent[v] = u;
				}
			}
		}
		
		/// Check for negative weight cycles
		for(int i=0; i<E; i++) {
			int u = edges.get(i).source;
			int v = edges.get(i).dest;
			int w = edges.get(i).weight;
			
			if (distance[u] != Integer.MAX_VALUE && (distance[u]+w) < distance[v]) {
				System.out.println("Negative weight cycle Found!!");
			}
		}
		
		for (int i=0; i<N; i++) {
			System.out.print("Distance of vertex " + i + " from the source is " + distance[i] + ". It's path is [ ");
			printPath(parent, i);
			System.out.println("]");
		}
		
	}
	
	public static void main(String[] args) {
		List<Edge> edges = Arrays.asList(
				new Edge(0, 1,-1),  new Edge(0, 2, 4),
				new Edge(1, 2, 3),  new Edge(1, 3, 2),
				new Edge(1, 4, 2),  new Edge(3, 2, 5),
				new Edge(3, 1, 1),  new Edge(4, 3,-3)
		);
		int source = 0;
		int N = 5;
		
		BellmanFord(edges, source, N);
	} 
}
